Add RFC for adaptiveMetadata table feature (#6978)
delta-io/delta: protocol_rfcs/iceberg-v4-metadata.md

# Iceberg V4 Adaptive Metadata Tree
**Associated Github issue for discussions: https://github.com/delta-io/delta/issues/6640**

> This RFC introduces a new reader-writer table feature `adaptiveMetadata` that enables Delta tables to adopt [Apache Iceberg™ V4's adaptive metadata tree](https://s.apache.org/iceberg-single-file-commit) as their native content metadata format.

Reviewer: nit
```suggestion
This RFC introduces a new reader-writer table feature `adaptiveMetadata` that enables Delta tables to adopt the [Apache Iceberg™ V4 adaptive metadata tree](https://s.apache.org/iceberg-single-file-commit) as their native content metadata format.
```

## Motivation

> Current Delta checkpoints rewrite the entire table state on every checkpoint, regardless of how much actually changed. This makes metadata costs proportional to table size rather than operation size, and prevents caching since checkpoint files change completely between versions.

Reviewer: Also: Today's Delta must store all changes of a commit directly in the commit .json file, regardless of how big the commit was. Gives up ~10x compression (json vs. parquet), impossible to parallelize, and causes high memory pressure at both read and write time.

Author: Great point. Added to motivation.

- Metadata changes proportional to operation size
- Stable leaf manifests that can be cached effectively

--------

# Changes to existing sections

### Add File

> ***Change to [existing section](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#add-file-and-remove-file)***

<ins>When the `adaptiveMetadata` table feature is enabled, the `add` action supports a `backReference` field:</ins>

| Field Name | Data Type | Description |
| - | - | - |
| <ins>backReference</ins> | <ins>Struct</ins> | <ins>Reference to the existing entry in the V4 metadata tree that this add supersedes (e.g., stats backfill, DV update). Null when the add is for a genuinely new file not already in the tree. Contains `manifest` (String) and `pos` (Long). See [Backreferences](#backreferences).</ins> |

### Remove File

> ***Change to [existing section](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#remove-file)***

<ins>When the `adaptiveMetadata` table feature is enabled, the `remove` action requires a `backReference` field and `deletionTimestamp` must be null:</ins>

| Field Name | Data Type | Description |
| - | - | - |

> | <ins>deletionTimestamp</ins> | <ins>Long</ins> | <ins>Must be null. Metadata cleanup uses tree reachability instead of timestamp-based expiration.</ins> |

Reviewer: Cross-check: Does the RFC update checkpoint spec to no longer include remove tombstones? Similarly, I don't think there's any reason for action reconciliation to include removes now. Scans never needed them, and now checkpoints don't need them either.

Reviewer: See also https://github.com/delta-io/delta/pull/6982#issuecomment-4662488236, for another potential wart involving removes and CRC. Tho perhaps that one is still relevant as some kind of safety check?

Author: Yes. I have updated the RFC to state this explicitly. Specifically:

- Checkpoints don't retain `remove` tombstones
- Action reconciliation no longer carries removes through. 

On the #6982 wart with `numRemoves`, I think it shouldn't cause issues for AMT-enabled tables? From what I could tell, there are two consumers of the remove count:

1. Multipart checkpoint sizing
2. Cheap snapshot reconstruction: numRemoves is not part of the hard table state, so do we need it?

Reviewer: I don't think `numRemoves` would be very useful or even relevant once we stop storing remove tombstones in checkpoints for VACUUM purposes. But I'm not confident that I understand what the CRC `numRemoves` field was actually trying to accomplish in the first place.

| <ins>backReference</ins> | <ins>Struct</ins> | <ins>Required reference to the file's location in the V4 metadata tree. Contains `manifest` (String) and `pos` (Long). See [Backreferences](#backreferences).</ins> |

### Last Checkpoint File

> ***Change to [existing section](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#last-checkpoint-file)***

<ins>The `_last_checkpoint` file remains a non-authoritative hint, as defined in the existing protocol. Readers discover whether a table supports `adaptiveMetadata` from the `protocol` action in the log or checkpoint, not from `_last_checkpoint`.</ins>

> <ins>When the `adaptiveMetadata` table feature is enabled and a manifest commit has been written, the `_last_checkpoint` file may contain the following additional fields:</ins>

Reviewer: Would it be better to just embed a checkpoint action in the _last_checkpoint file, rather than (partially?) replicating the definition here by a flat collection of fields? That way, we can centralize the explanation of all fields and the action's presence authoritatively identifies an adaptiveMetadata table. Absence means you have to actually do log replay to be sure, unless we want to keep the `v4` flag as well. That said, I don't know why a writer would ever _not_ emit the checkpoint action when checkpointing a v4 table?

Author: Yes, that makes sense. `_last_checkpoint` now embeds the `checkpoint` action directly instead of replicating the fields. Also dropped the `v4` flag.

| Field Name | Data Type | Description |
| - | - | - |

> | <ins>v4</ins> | <ins>Boolean</ins> | <ins>If true, indicates this is a V4 checkpoint.</ins> |

Reviewer: This "v4" thing is a bit confusing. Does it refer to a Delta checkpoint protocol version or the v4 iceberg spec? And Iceberg V4 includes more than just the adaptive metadata tree (AMT). Should we refer to `adaptiveMetadata` (table feature) or AMT (data structure) instead of "v4"? Or at least specify clearly up front what this RFC means when it says "v4" ?

Author: I added a `Terminology` section. I also took a pass the the rest of the RFC and cleaned up.

> | <ins>contentRoot</ins> | <ins>Struct</ins> | <ins>Reference to the V4 root manifest (path, sizeInBytes).</ins> |

Reviewer: Other Delta file references (add, remove, etc) include modification time (**). Should we do the same here?

(**) As a nod to hadoop-based readers, who will issue a HEAD request under the hood if that information is missing.

Author: `modTime` is not supported in the Iceberg v4 manifest files, so we were considering dropping it. cc @emkornfield

| <ins>protocol</ins> | <ins>Struct</ins> | <ins>The Protocol action for this checkpoint version.</ins> |
| <ins>metadata</ins> | <ins>Struct</ins> | <ins>The Metadata action for this checkpoint version.</ins> |

> | <ins>domainMetadata</ins> | <ins>Array[Struct]</ins> | <ins>Optional. DomainMetadata actions at this checkpoint version. Presence is authoritative; absence means the reader must consult the checkpoint action or sidecars.</ins> |

Reviewer: Two things:
1. What does "authoritative" mean? (I believe it means "the list is complete, if present")
2. At one point we discussed always including system domains (few in number, and important) while only including unbounded user domains on a best-effort basis?

Author: 1. Correct. 
2. I've added the notion of a system vs user domain. System domains are always inline while user domain could be inlined (if small) or in a sidecar. (following @prakharjain09 's suggestion [here](https://github.com/delta-io/delta/pull/6978/changes#r3471142776).

> | <ins>txns</ins> | <ins>Array[Struct]</ins> | <ins>Optional list of SetTransaction actions. Each entry contains appId, version, and optional lastUpdated. When small, txns are embedded here; when large, stored in sidecars.</ins> |

Reviewer: ```suggestion
| <ins>txns</ins> | <ins>Array[Struct]</ins> | <ins>Optional list of SetTransaction actions. Each entry contains appId, version, and optional lastUpdated. When the list is small, it is embedded here; large lists are stored in sidecars instead.</ins> |
```

| <ins>sidecars</ins> | <ins>Array[Struct]</ins> | <ins>Optional list of sidecar entries for overflow txns and domain metadata only. File-level metadata is stored in the content tree. Each entry contains path, sizeInBytes, modificationTime, and tags (matching the existing sidecar action schema).</ins> |

> <ins>When `v4` is true, readers can use `contentRoot.path` to begin prefetching the root manifest immediately. The `protocol`, `metadata`, `domainMetadata`, `txns`, and `sidecars` fields provide the complete table state at `version`. If the hint is absent, stale, or does not contain V4 fields, readers fall back to log replay to locate the latest `checkpoint` action.</ins>

Reviewer: Does `v4: true` imply that all other fields must be present, except the domain metadata and txn lists that may be omitted because too large?

Author: That is accurate. BTW I have removed the v4 flag: _last_checkpoint doesn't carry it anymore.

### Checkpoints

> ***Change to [existing section](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#checkpoints)***

<ins>When the `adaptiveMetadata` table feature is enabled, checkpoint information may be embedded directly in Delta log entries via the `checkpoint` action, rather than stored in separate checkpoint files. See [Checkpoint Action](#checkpoint-action) for details.</ins>

<ins>The `checkpoint` action is self-contained: it embeds the content root reference along with all non-content metadata (`protocol`, `metaData`, `domainMetadata`, `txns`, `sidecars`). Together with the referenced manifest tree, a single `checkpoint` action represents the complete table state up to `checkpoint.version`. Log commits after that version must still be replayed.</ins>

> <ins>Traditional checkpoints (single-file, multi-part, and V2) remain supported. Writers may produce a traditional checkpoint when: the table is being read by clients that do not support `adaptiveMetadata` during a rolling upgrade, or when the feature is being removed and the final checkpoint must be readable without the V4 tree.</ins>

Reviewer: This section is confusing to me:
1. Should we point out that a normal v4 checkpoint is just a `checkpoint` action stored separately, because the metadata tree holds all content metadata as usual?
2. It makes sense that a tree would still contain traditional checkpoints for versions that predate an upgrade. But it's _very_ unclear to me how log replay could safely use such a checkpoint (**)?  

(**) A major benefit of AMT is that separate checkpoint files are no longer necessary for correctness (because manifest commits embed all the same information directly). So it would be weird for a v4 tree to take a dependency on a side file that could be deleted at any time for any reason. Instead, I would expect the upgrade to be a manifest commit. For all but the smallest tables, that commit would introduce a new tree root to capture content metadata. For tiny tables, the manifest commit could directly embed the entire table state as a rollup. Either way, log replay would completely ignore all legacy checkpoints before the upgrade, because log replay always ignores all commits before a manifest commit.

By the same reasoning, a downgrade would need to physically back out the upgrade by making a new manifest commit with a null tree root (***), and that commit would then have to roll up _all_ table state. That potentially _massive_ commit would naturally trigger creation of a traditional checkpoint immediately after the commit lands. But if we're anyway making that massive commit, why not just remove the feature entirely at the same time? Is the multi-step process needed in order to leverage [checkpoint protection](https://github.com/delta-io/delta/blob/master/protocol_rfcs/checkpoint-protection.md) for a smoother downgrade experience?

Methinks we'll need to be _very_ careful in designing the upgrade and downgrade paths, if we want to support them at all. 

(***) Do we actually have a way to discard an existing tree without replacing it, e.g. if a table shrinks (or we're prepping for a downgrade)?

Reviewer: Related: AFAIK, we have two technically orthogonal axes for a commit:
1. Does it introduce a new `checkpoint` action? (checkpoint commit)
2. Does it introduce a new metadata tree root? (manifest commit)

For simplicity, we require that every manifest commit must also be a checkpoint commit, tho we do allow checkpoint commits that are not manifest commits (i.e. to roll up deltas that accumulated since the tree was last modified).

And of course we allow commits that are neither checkpoint nor manifest commits. Indeed, those would be the common case for workloads dominated by small DML commits.

But is "non-manifest checkpoint commit" _really_ a helpful distinction?
* The word "checkpoint" has a _LOT_ of baggage in Delta, and we're adding a _LOT_ of potential ambiguity here:
   * "checkpoint commit" (noun)
   * "checkpoint a commit" (verb, quite similar to the noun but actually completely different)
   * "commit a checkpoint" (verb, corresponds to the noun)
   * "checkpoint" (noun, referring to the set of files comprising a checkpoint for a given commit)
   * "checkpoint" (noun, sloppy shorthand for "checkpoint commit")
   * "checkpoint" (verb, sloppy shorthand for "checkpoint a commit")
   * "checkpoint" (verb, sloppy shorthand for "commit a checkpoint")
* If we anyway need to support manifest commits that remove an existing tree root without replacing it (for downgrade), why not allow "no-op" manifest commits that re-state an existing tree root as well?
* That way, the "checkpoint commit" and "manifest commit" concept would collapse into just "manifest commit" (which may drop the current tree root, replace it, or re-state it). 
   * Leaving only two kinds of commits: manifest commits (which include all non-content metadata and roll up all content metadata after the tree root) and normal (delta-only) commits.
   * The `checkpoint` action would be better named a `manifest` action or similar
* And then "checkpoint" could keep its current meaning: An auxiliary file created alongside a commit, which rolls up all table state and can be used to directly create a snapshot without log replay. It just so happens that a "checkpoint" of an `adaptiveMetadata` table will ~always be a single small file because the metadata tree still holds the bulk of table state.

Author: We don't use the word checkpoint commit anywhere. The model is that we have manifest commits which produce checkpoints, and checkpoints embeds a content metadata tree.  I think it makes sense to call the action a checkpoint action because it encompasses content metadata and non-content metadata of the table.  So we are retaining the current semantics of checkpoints, which is an encapsulation of the complete table state. 

I do agree that it is better to make the reader-writer feature mutually exclusive with traditional checkpoints. I also made the change such that an upgrade to the adaptive metadata tree feature produces a manifest commit so that there is no dependency on previous traditional checkpoints. 

About the null root my inclination is to have an empty content root with no files inside. Or if the table has a small number of files, have a small root on disk. This would help us avoid having special branches in the code to handle the null content root, and also we have consistency with Iceberg which always will have a root on disk.

### Deletion Vectors

> ***Change to [existing section](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#deletion-vectors)***

> <ins>When the `adaptiveMetadata` table feature is enabled, inline deletion vectors (storage type `i`) are not supported. The V4 content entry represents deletion vectors as file references (`location`, `offset`, `size_in_bytes`, `cardinality`) with no field for inline bytes. Storage types `u` (UUID-relative) and `p` (absolute path) are supported since both resolve to file paths. When writing V4 manifest entries, writers must resolve the `u` encoding to a relative path (e.g., `data/deletion_vector_<uuid>.bin`) for the DV `location` field. Existing inline DVs must be converted to file-based DVs before or during feature enablement.</ins>

Reviewer: ```suggestion
<ins>When the `adaptiveMetadata` table feature is enabled, inline deletion vectors (storage type `i`) are forbidden. The V4 content entry represents deletion vectors as file references (`location`, `offset`, `size_in_bytes`, `cardinality`) with no field for inline bytes. Storage types `u` (UUID-relative) and `p` (absolute path) are supported since both resolve to file paths. When writing V4 manifest entries, writers must resolve the `u` encoding to a relative path (e.g., `data/deletion_vector_<uuid>.bin`) for the DV `location` field. Existing inline DVs must be converted to file-based DVs before or during feature enablement.</ins>
```

Reviewer: Also -- this seems to conflate the v4 format (which AFAIK only supports one kind of DV descriptor) with upgrade procedure (how to convert existing file actions to v4 format)?

Reviewer: Tho I guess we also need to do the same conversion when flushing deltas down into the metadata tree?

--------

> ***Add a new section at the [Table Features](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#table-features) section***

# Adaptive Metadata

The `adaptiveMetadata` table feature enables Delta tables to store table state using an adaptive metadata tree structure. When enabled, commits can choose between writing changes to the Delta log (*log commits*) or producing a V4 metadata tree with an embedded checkpoint action (*manifest commits*).

## Table Feature Enablement

The `adaptiveMetadata` table feature is supported when:
- The table is on Reader Version 3 and Writer Version 7.
- The feature `adaptiveMetadata` exists in the table `protocol`'s `readerFeatures` and
  `writerFeatures`.

Required table features that must also be enabled:
- `columnMapping` (`id` mode): Stable column identification
  across schema evolution
- `rowTracking`: Deletion vector rebasing during conflict
  resolution
- `domainMetadata`: Storing V4-specific metadata
- `deletionVectors`: Row-level conflict resolution during
  concurrent writes
- `inCommitTimestamp`: Reliable timestamp-based metadata

>   cleanup and time travel

Reviewer: nit: Weirdly short line breaks?

Author: Fixed.

## Storage Layout

> When `adaptiveMetadata` is enabled, the table storage layout includes a `metadata/` directory alongside the standard `_delta_log/`:

Reviewer: Should it be `_metadata` (hidden dir, according to Delta spec)? Otherwise technically users could put data files in there and/or VACUUM could run amok (according to current Delta spec).

Author: We _could_, but it doesn't protect us from Iceberg writers. In  Iceberg, by default data files are written to to `<table>/data` and metadata files to `<table>/metadata`, but both are overridable.

We need to make VACUUM and metadata cleanup reachability-based. ie never delete a data file, DV, or manifest that's reachable from the live tree or retained log, whatever directory it's in. I'll update the Metadata Cleanup section accordingly. 

Given that, I was thinking that we should just be consistent with Iceberg conventions.

```
table/
├── _delta_log/
│   ├── 00000000000000000042.json     # Commit with embedded checkpoint action

> │   ├── 00000000000000000043.json     # Log commit (small changes only)

Reviewer: Those comments are a bit weird -- there's no way to "see" what they claim. Does this correspond to some kind of running example that the text will elaborate?

Author: I was using the comments as an illustration. I can remove it if they are confusing.

The `metadata/` directory is the default and recommended location for V4 manifests. Manifest paths in the protocol are stored as **relative paths from the table root** (e.g., `metadata/root-v42.parquet`). A path is relative if it does not contain a URI scheme. Relative paths are resolved by joining the table location and the relative path with a `/` separator (`table_location + "/" + relative_path`). Absolute paths (with a URI scheme) are used as-is. Relative paths must not start with `/` and table locations must not end with `/` to avoid duplicate separators. This follows the [Iceberg V4 relative paths specification](https://iceberg.apache.org/spec/#paths-in-metadata).

The `_delta_log/_sidecars/` directory stores auxiliary data that is too large to embed in the checkpoint action.

## Checkpoint Action

When a manifest commit occurs, the Delta log entry contains a self-contained `checkpoint` action that includes all non-content metadata needed to reconstruct the full table state.

```json
{
  "checkpoint": {
    "version": 42,
    "contentRoot": {
      "path": "metadata/a3d1f7e2-v42.parquet",
      "sizeInBytes": 1024

>     },

Reviewer: The `contentRoot` needs to specify the commit version it corresponds to, no?
* IIUC, it's legal for a writer to produce a new content root for an older version, as long as it's newer than the previous content root's version
* The previous section specifically states that "The protocol does not prescribe manifest naming conventions" so we can't rely on parsing the root's file name.

Unless the top-level `version` actually refers to the content root, rather than the checkpoint itself? Or we're not worried about enforcing the requirement that roots progress monotonically through the table history?

Author: My initial thinking that `contentRoot` and checkpoint versions should be the same. So if a writer is producing a content root for an older version of the tree, the rest of the checkpoint would be for that version.  

But rethinking, maybe we want to allow them to be different for standalone checkpoints. But we probably should not allow them to be different for manifest commits?  Otherwise it would bloat the JSON log files with minor-compacted-entries that should have been in the content tree instead?

Reviewer: Interesting. So you're saying that a manifest commit should checkpoint the commit itself, even if the new content root is for an older commit, and just roll up _all_ deltas after the root. Meaning that the entire content of the commit is just a checkpoint? There is a certain appeal to that (avoids needing to replay a bunch of older commits). 

I _think_ it would also mean that we no longer need a specific `checkpoint` action? If the commit contains a `contentRoot` action, then we need look no further -- any such commit is _required_ to be a rollup and it would be _incorrect_ to process the intervening commits?

Author: I was saying that a manifest commit `N` checkpoints _a_ version `C` where `C` is <= `N`. `C` becomes `checkpoint.version` and `contentRoot.version`

We don't always want to force a manifest commit to checkpoint its own version because it will increase the chance of conflicts (e.g. with other log commits). But in some cases, it is desirable to checkpoint its own version. e.g. the commit itself is a very large change, and it's cheaper to make changes to the metadata tree rather than serially write a gigantic JSON commit file. 

Given that checkpoint version could be older than the manifest commit version, we can't use `contentRoot`s presence as the barrier. Reader must still replay commits after `checkpoint.version`, including the manifest commit that produced the checkpoint.

    "protocol": {
      "minReaderVersion": 3,
      "minWriterVersion": 7,
      "readerFeatures": ["columnMapping", "deletionVectors", "adaptiveMetadata"],
      "writerFeatures": ["columnMapping", "deletionVectors", "domainMetadata",

>                          "rowTracking", "adaptiveMetadata"]

Reviewer: Another weird line break?

Author: Fixed.

    },
    "metaData": {
      "id": "af23c9d7-fff1-4a5a-a2c8-55c59bd782aa",
      "name": "my_table",
      "schemaString": "{...}",
      "partitionColumns": [],
      "configuration": {},
      "createdTime": 1234567890000
    },
    "domainMetadata": [
      {
        "domain": "delta.rowTracking",
        "configuration": "{\"rowIdHighWaterMark\": 1000000}",
        "removed": false
      }
    ],
    "txns": [
      {
        "appId": "streaming-app-1",
        "version": 100,
        "lastUpdated": 1234567890000
      }
    ],
    "sidecars": [
      {
        "path": "txn-v42.parquet",
        "sizeInBytes": 2048,
        "modificationTime": 1234567890000,
        "tags": {}
      }
    ]
  }
}
```

### Checkpoint Action Fields

| Field | Type | Description |
|-------|------|-------------|

> | `version` | Long | The table version up to which the checkpoint is complete. May be less than or equal to the commit version (e.g., commit v100 may checkpoint v50). Checkpoint versions must be strictly monotonically increasing across all checkpoint actions in the log (the next checkpoint must be for a version > 50). |

Reviewer: Ah, so maybe we don't need to make any requirements about tree versions? If the checkpoint actions progress monotonically through table history, then it doesn't matter which version given tree corresponds to, because the checkpoint anyway must have rolled up all deltas that landed since then?

Author: I think the content root version is the one whose monotonicity that matters (to make sure that the back references never go stale). I have updated it. PTAL.

| `contentRoot` | Struct | Reference to the V4 root manifest file |
| `contentRoot.path` | String | Path to the root manifest, relative to the table root. May be an absolute URI for files outside the table root. |
| `contentRoot.sizeInBytes` | Long | Size of the root manifest file in bytes |
| `protocol` | Struct | The Protocol action at this checkpoint version |
| `metaData` | Struct | The Metadata action at this checkpoint version |
| `domainMetadata` | Array[Struct] | Optional. DomainMetadata actions at this checkpoint version |
| `txns` | Array[Struct] | Optional. SetTransaction actions (appId, version, lastUpdated). When large, stored in sidecars. |
| `sidecars` | Array[Struct] | Optional. Sidecar file references for overflow txns and domain metadata only. File-level metadata is stored in the content tree. Schema: path, sizeInBytes, modificationTime, tags. |

> An `appId` must not appear in both the inline `txns` array and a sidecar file. The complete set of transaction identifiers at `checkpoint.version` is the union of inline `txns` and sidecar contents, with no duplicates across the two.

Reviewer: So, unlike the claims from _last_checkpoint, this is _not_ "authoritative" ? 
Or it's "authoritative" but only on a per-appid basis?

## Backreferences

> When `adaptiveMetadata` is enabled, `remove` and `add` actions carry a `backReference` field that identifies where the file's existing entry is located in the metadata tree. Backreferences are required when the file has an entry in a manifest. They are null when the file only exists in the Delta log (not yet incorporated into a manifest) or is genuinely new (first add of a path).

Reviewer: Is "genuinely new" really a relevant case, let alone a separate one? A genuinely new file could be incorporated directly into a new manifest, and files that only exist in the Delta log may not be new.

Author: That is actually true - what matters is not whether the file is genuinely known, but whether it exists in a manifest. I have reworded this.

Reviewer: The text still says 
> They are null when the file only exists in the Delta log (not yet incorporated into a manifest) or is genuinely new (first add of a path).

... which is the incorrect idea that a "new" file is somehow special. Backreferences are NULL if the file is log-only, and non-NULL if the file is live in a manifest. What does new/old have to do with anything?

Author: I forgot to update the text. Done. PTAL.

> Backreferences enable efficient MDV creation during manifest commits: writers can directly construct MDVs from backreferences without scanning leaf manifests. The engine must propagate (manifest, position) metadata through the planning pipeline so it is available at commit time.

Reviewer: Probably worth defining the MDV acronym here? Any intro matter that defines it is either part of the RFC motivation (not in final spec) or in the feature intro (far away from here).

Reviewer: At least link to the MDV section below.

Author: Done. Now expands it to "Manifest Deletion Vector (MDV)" and links to the MDV section.

### Remove with Backreference

```json
{
  "remove": {
    "path": "data/part-00001.parquet",
    "dataChange": true,
    "deletionVector": {
      "storageType": "u",
      "pathOrInlineDv": "ab^-aqEH",
      "offset": 4,
      "sizeInBytes": 40,
      "cardinality": 6
    },
    "backReference": {
      "manifest": "metadata/leaf-m1.parquet",
      "pos": 17
    }
  }
}
```

### Add with Backreference (Re-add)

> When an `add` supersedes an existing entry (e.g., `OPTIMIZE` backfilling stats on a file), the backreference points to the old entry:

Reviewer: ```suggestion
When an `add` supersedes an existing manifest entry (e.g., `OPTIMIZE` backfilling stats on a file), the backreference points to the old entry:
```

```json
{
  "add": {
    "path": "data/part-00001.parquet",
    "size": 1234567,
    "stats": "{\"numRecords\":1000,\"minValues\":{\"id\":1},\"maxValues\":{\"id\":1000}}",
    "backReference": {
      "manifest": "metadata/leaf-m1.parquet",
      "pos": 17
    }
  }
}
```

### Backreference Fields

| Field | Type | Description |
|-------|------|-------------|
| `manifest` | String | Path to the leaf manifest containing this file, relative to the table root (e.g., `metadata/leaf-m1.parquet`) |
| `pos` | Long | Row position (0-indexed) of the file entry within the manifest |

## V4 Content Entry Schema

> Both root and leaf manifests use a single unified entry schema, following the [Iceberg V4 content entry design](https://github.com/apache/iceberg/pull/15049). Certain fields are only applicable to specific `content_type` values (noted below). Manifests are Parquet files; **projection must be done by field ID, not by name.** Writers must set Parquet `field_id` metadata on all fields so that readers can resolve them by field ID.

Reviewer: ```suggestion
Both root and leaf manifests use a single unified entry schema, following the [Iceberg V4 content entry design](https://github.com/apache/iceberg/pull/15049). Certain fields are only applicable to specific `content_type` values (noted below). Manifests are Parquet files. Writers must set Parquet `field_id` metadata on all fields so that readers can resolve them by field ID, and **readers must resolve fields by field ID, not by name.**
```

Leaf manifests contain entries for data files only (`content_type` = DATA). They cannot reference other manifests: this enforces a two-level tree hierarchy (root -> leaves only).

> The root manifest contains entries of the following types:

Reviewer: IIRC, Iceberg has a notion of delete manifests in a snapshot, which record files that were removed by the commit which produced the snapshot (mostly informational, used for change data capture, perhaps)? Is that just not a thing in Iceberg v4? Or is this proposal adding an additional restriction to not emit such manifests?

Author: In Iceberg V3, we have a delete manifest for keeping track of deletion vectors that need to be joined with data files during reads. This is not just used for CDF. In V4, we are moving towards a co-located data manifest where one row contains the data file as well as a DV. We don't need delete manifest anymore. The only exception for this is equality deletes which is not supported in delta.

There is another case where the Iceberg table can be upgraded from existing v3 tables. We probably will not support this scenario, as the v3 Avro manifest is not supported in the unified metadata.

### Tracking

| Field ID | Field Name | Delta Type | Required | Description |
|----------|------------|------------|----------|-------------|
| 0 | `status` | Int | Required | 0=EXISTING, 1=ADDED, 2=DELETED, 3=REPLACED |
| 1 | `snapshot_id` | Long | Optional | Snapshot ID where entry was added or deleted. Inherited when null. |
| 3 | `sequence_number` | Long | Optional | Data sequence number. Inherited when null and status=ADDED. |
| 4 | `file_sequence_number` | Long | Optional | File sequence number (when file was physically added). Inherited when null and status=ADDED. |
| 5 | `dv_snapshot_id` | Long | Optional | Snapshot ID where DV was added. Null when no DV. |
| 142 | `first_row_id` | Long | Optional | Starting row ID for this file (DATA) or manifest (DATA_MANIFEST) |
| 6 | `deleted_positions` | Binary | Optional | Bitmap of positions deleted in this commit (DATA_MANIFEST only, for CDF) |
| 7 | `replaced_positions` | Binary | Optional | Bitmap of positions replaced in this commit (DATA_MANIFEST only, for CDF) |

**Status values:** `status` is required and always materialized (never null). Each entry has exactly one status.

> Only `EXISTING` and `ADDED` entries are live (visible to scans). `DELETED` and `REPLACED` entries are not live and exist only for change data feed (CDF) reconstruction. A file path has at most one live entry in the tree at any time.

Author: Note to self: add the `MODIFIED` status when it's added to Iceberg spec.

> Only `EXISTING` and `ADDED` entries are live (visible to scans). `DELETED` and `REPLACED` entries are not live and exist only for change data feed (CDF) reconstruction. A file path has at most one live entry in the tree at any time.

Reviewer: How should a Delta table interpret these informational DELETED/REPLACED entries when reading a tree (that was possibly created for a previous version) or when creating a new tree? I _think_ they're truly dead bloat? In order to be useful for CDF reconstruction, we'd need to know which table version the tree corresponds to, since a `checkpoint` action is not required to create a new tree and therefore the `checkpoint.version` field is not trustworthy?

Author: Yes, they are dead bloat, but these entries drop off in the next version of the tree. Yes, we just added the version of the metadata tree, we have a trustworthy field that indicates which version of the table corresponds to a content metadata tree.

- `EXISTING` (0): Unchanged from previous commit. Live.
- `ADDED` (1): New in this commit. Live.
- `DELETED` (2): Removed in this commit. Not live.
- `REPLACED` (3): Superseded (e.g., DV added/changed). Not live. A
  new DATA entry with updated metadata exists elsewhere in the tree as the live entry for this file path.

Non-live entries are transient: they are dropped in the next manifest commit that rewrites the affected manifest (root or leaf).

> **Inheritance:** Inheritance applies only to leaf manifest entries inheriting from the root. Following the [Iceberg specification](https://iceberg.apache.org/spec/#sequence-number-inheritance):

Reviewer: Naive question: What other inheritance types exist?
And why would we not want to support those other types? 

For example, I believe row tracking information could be inherited from manifest through a leaf through a data file all the way to individual rows, which would make it a lot easier to rebase row tracking information during a commit retry (no need to rewrite newly-added leaf manifests)?

Reviewer: Ah, that specific row tracking scenario is literally the next thing below.
But the original question still stands?

Author: > What other inheritance types exist? And why would we not want to support those other types?

There is only one kind of inheritance mechanism. That is, you can omit certain fields from leaf manifest files, and those values will be filled in from the root manifest. This makes retries light-weight. 

The inheritable fields are the four Iceberg defines  are `snapshot_id`, `sequence_number`, `file_sequence_number`, `first_row_id`,  and we inherit all of them. 

We have a separate section on road tracking compatibility. Please let me know if this is clear enough.

1. A writer creates a leaf manifest with `ADDED` entries. Tracking
   fields like `sequence_number` are null because the commit version is not yet assigned.
2. The writer creates the root manifest with a `DATA_MANIFEST` entry
   referencing the leaf. The `DATA_MANIFEST` entry has explicit (non-null) tracking values assigned at commit time.
3. A reader encounters a null tracking field in a leaf entry and
   inherits the value from the `DATA_MANIFEST` entry in the root that references that leaf.

The fields that support inheritance:

- **`snapshot_id`**: Inherited when null.
- **`sequence_number`**: Inherited when null and status is `ADDED`.

> - **`file_sequence_number`**: Inherited when null and status is `ADDED`.

Reviewer: These must be non-null when EXISTING?

Author: Yes. Inheritance fills nulls only for ADDED leaf entries, so any non-ADDED entry (EXISTING/DELETED/REPLACED) must carry explicit, non-null values for all inheritable fields.

### Manifest Info

Summary information for `content_type` = DATA_MANIFEST entries. Includes file/row counts and the Manifest Deletion Vector (MDV).

| Field ID | Field Name | Delta Type | Required | Description |
|----------|------------|------------|----------|-------------|
| 504 | `added_files_count` | Int | Required | Number of files added |
| 505 | `existing_files_count` | Int | Required | Number of existing files |
| 506 | `deleted_files_count` | Int | Required | Number of deleted files |
| 520 | `replaced_files_count` | Int | Required | Number of replaced files |
| 512 | `added_rows_count` | Long | Required | Number of rows in added files |
| 513 | `existing_rows_count` | Long | Required | Number of rows in existing files |
| 514 | `deleted_rows_count` | Long | Required | Number of rows in deleted files |
| 521 | `replaced_rows_count` | Long | Required | Number of rows in replaced files |
| 516 | `min_sequence_number` | Long | Required | Minimum sequence number of files in this manifest |
| 522 | `dv` | Binary | Optional | MDV bitmap marking deleted positions in leaf manifest. Must be non-null if and only if `dv_cardinality` is non-null. |

> | 523 | `dv_cardinality` | Long | Optional | Number of entries marked as deleted in the MDV. Must be non-null if and only if `dv` is non-null. |

Reviewer: Simpler to say they must be null or non-null together?

Author: Done.

### Manifest Deletion Vectors (MDVs)

> MDVs track which entries in a leaf manifest have been deleted or replaced. The `dv` field in `manifest_info` contains a Roaring bitmap of all invalidated positions (cumulative, used for reading).  The `deleted_positions` and `replaced_positions` fields in `tracking` represent what changed in the current commit only (used for CDF):

Author: The MDV bitmap is currently under-specified intentionally. It is possible that Iceberg will adopt the [Mumbling](https://github.com/apache/iceberg/pull/16518) bitmap proposed by @rdblue.

Reviewer 2: How would existing DV's be handled, would readers need to handle the old `.bin` format as well as the new definition or will there be a need to migrate data to the new version ?

Reviewer: AFAIK, the ambiguity only applies to MDV, which don't yet exist? The linked draft spec says "for embedded deletion vectors in v4 root manifests" so normal file DV (which are referenced by leaf manifests) would continue to use roaring bitmaps either way?

Reviewer: But meanwhile it seems only partly under-specified?
* `manifest_info.dv` "contains a Roaring bitmap"
* `tracking.deleted_positions` and `tracking.replaced_positions` technically don't specify a format

Author: > But meanwhile it seems only partly under-specified?

That is correct. I mention roaring bitmap once, but will clarify it when the decision to use Mumbling is made in the Iceberg community. 

@jayadeep-jayaraman The metadata DVs is always stored inline in the root manifest (not in separate bin files). The format is TBD. (could either be Roaring or the new Mumbling format being proposed).

Reviewer 2: Thanks @anoopj & @scovich

- **`manifest_info.dv`**: Cumulative set of all invalidated positions. Grows
  monotonically. Used by readers to skip entries.
- **`tracking.deleted_positions`**: Positions newly deleted in this
  commit (for CDF).
- **`tracking.replaced_positions`**: Positions newly replaced in this
  commit (for CDF).

A position cannot be set in both `deleted_positions` and `replaced_positions` simultaneously.

When a `DATA_MANIFEST` entry has a non-null MDV, its `content_stats` are stale: aggregate bounds and counts still include values from invalidated entries. Writers must set `tight_bounds = false` on the manifest-level stats when an MDV is present. After compaction clears the MDV, stats are recomputed from live entries only and `tight_bounds` can be true again.

### Content Stats

The `content_stats` field (ID 146) uses Iceberg V4's typed column statistics structure, as defined in the [Iceberg V4 specification](https://iceberg.apache.org/spec/#content-stats).

> The `content_stats` struct is optional on the entry, and each per-column stats struct within it is also optional. Writers are not required to produce stats for every column.

Reviewer: What about stats in deltas? Today's Delta spec is very ambiguous on what requirements, if any, exist for stats on `remove` actions, which can complicate operations such as log replay, merge, and transaction conflict retry. Should we tighten up the requirements so that e.g. a `remove` must propagate whatever stats the corresponding `add` provided?

Author: Currently the protocol has remove.stats as an optional field, and it only lives in JSON commits. Is the suggestion to make it required, just in the JSON commits? Did you have a specific use case in mind? This is technically not needed for AMT, but we can consider making it _required_ if needed

Related: We are making the behavior of `extendedFileMetadata` default with v4.  (which carries forward partitionValues, size, but not stats).

Author: OK, I made remove.stats as required. PTAL

Only leaf (primitive) columns get stats structs. Complex types (struct, list, map) do not. The stats structs are flat fields in `content_stats`, not nested under parent types. Field IDs are deterministically calculated from table field IDs:

| Table Field ID | Stats Struct Field ID | Calculation |
|----------------|----------------------|-------------|
| 0 | 10000 | 10_000 + 200 × 0 |
| 1 | 10200 | 10_000 + 200 × 1 |
| 2 | 10400 | 10_000 + 200 × 2 |
| 100 | 30000 | 10_000 + 200 × 100 |

Each per-column stats struct contains:

| Offset | Field Name | Delta Type | Included For | Description |
|--------|------------|------------|--------------|-------------|
| +1 | `lower_bound` | (typed) | all primitives | Lower bound, type matches the table column type |

> | +2 | `upper_bound` | (typed) | all primitives | Upper bound, type matches the table column type |

Reviewer: These are NULL only if the column was all-null, right?  So we can distinguish between "stat exists and is NULL" vs. "stat is NULL because missing" by comparing `value_count` vs. `null_value_count`?

Author: Yes. When the column is all null, `value_count` and `null_value_count` would be the same.  The bounds are null when either the column is all null or the stats just wasn't collected.

| +3 | `tight_bounds` | Boolean | all primitives | When true, bounds are exact min/max |

> | +4 | `value_count` | Long | all | Number of values (including nulls and NaN) |

Reviewer: This differs from the Delta stats tracking spec, which tracks record count at file level, not column level. 
Why would we want to replicate the value count for every column? 

Does iceberg collect stats on leaf columns nested inside arrays, so the count could vary for (leaf) column to (leaf) column? 

In any case, we probably need RFC language to reconcile `add.stats` in a commit file with manifest stats.

Author: We are not replacing the file level count. Iceberg also has `record_count` which is identical to the Delta equivalent. This RFC already adds it (field# 103)

> Does iceberg collect stats on leaf columns nested inside arrays, so the count could vary for (leaf) column to (leaf) column?

Yes, according to the spec. The reference implementation doesn't collect stats for repeated fields though. So this is not really used in practice. 

> In any case, we probably need RFC language to reconcile add.stats in a commit file with manifest stats.

Yes, I just added a "Stats Reconciliation" section to cover this. PTAL

| +5 | `null_value_count` | Long | optional fields | Number of null values |
| +6 | `nan_value_count` | Long | float, double | Number of NaN values |
| +7 | `avg_value_size_in_bytes` | Int | string, binary | Avg uncompressed value size in bytes |

> Implementations are not required to write a stats struct for every table field. If any field is missing from the struct, readers must assume it is unknown.

Reviewer: What about present-but-NULL, which is technically different but indistinguishable from how most parquet readers would interpret a physically missing column? Do readers need to actually examine individual file schemas from the parquet footer before deciding how to handle NULL stats?

Author: Footer inspection is not needed. All-null case can be inferred from the counts: `value_count` == `null_value_count`. Null upper/lower bound doesn't distinguish between all null vs missing stats.

**Aggregation rules** (for manifest-level stats in root manifest `DATA_MANIFEST` entries, computed over constituent leaf entries):

If any child entry is missing a stats field, the aggregate for that field must be null. Aggregation is only valid when all children have the field present.

- `value_count`, `null_value_count`, `nan_value_count`: sum
- `lower_bound`: min of non-null child values
- `upper_bound`: max of non-null child values
- `tight_bounds`: AND (false if any constituent is false)
- `avg_value_size_in_bytes`: must be null at the manifest level

## Snapshot ID Generation

The `snapshot_id` field in tracking identifies when content was added or modified. Writers must generate a unique long value for each manifest commit.

## Row Tracking Compatibility

Delta's row tracking fields map to Iceberg V4 tracking as follows:

| Delta Field | V4 Tracking Field | Field ID |
|-------------|-------------------|----------|
| `baseRowId` | `first_row_id` | 142 |
| `defaultRowCommitVersion` | `sequence_number`, `file_sequence_number` | 3, 4 |

> When `adaptiveMetadata` is enabled, Iceberg's `sequence_number` (data sequence number) and `file_sequence_number` are both set to the Delta commit version of the `add` action that introduced the file. They always resolve to the same value because Delta does not support late-arriving data (where a file's data logically belongs to an older version). `file_sequence_number` is required by Iceberg's inheritance model but Delta does not read it back.

Reviewer: > Delta does not support late-arriving data

I thought file_sequence_number != sequence_number was used to identify rewritten files that contain old data (e.g. COMPACT or OPTIMIZE)? Delta definitely supports those sorts of operations, do we just not care for some reason? Or do we still get the same information by detecting that sequence numbers differ between a manifest and (some of) the files that manifest contains, and/or differ between a file and (some of) the rows that file contains? But then for e.g. OPTIMIZE that combined rows from lots of input files to one output file, we no longer have a file-level upper bound and have to look at individual rows to figure it out?

Author: The reason Iceberg keeps a separate data sequence number is delete-file application (especially equality deletes apply to a data file by relative age). Delta binds DVs directly to their data file, so there's no sequence-ordering to resolve. 

In Iceberg, there are two use cases for file_sequence_number != sequence_number:

1. Late arriving data.  
2. OPTIMIZE producing files that logically belonging to an older table version. 

I will reword the RFC to remove the reference to the late arriving data rationale.

For ADDED entries in leaf manifests, both are null and inherited from the `DATA_MANIFEST` entry in the root (see [Inheritance](#inheritance)). For EXISTING entries (e.g., after compaction), both are materialized.

On compaction, the output file is a new physical file at the compaction commit version, so all three file-level fields (`defaultRowCommitVersion`, `sequence_number`, `file_sequence_number`) are the compaction version. The original per-row values (`_row_id`, `_row_commit_version`) are materialized as data columns in the output Parquet file to preserve row-level tracking.

## Partition Specs

Iceberg uses **partition specs** to define how a table is partitioned. Each partition spec has a unique `spec_id` and describes the partition columns.

Delta defines partitioning via `partitionColumns` in the table metadata and does not support partition evolution. When `adaptiveMetadata` is enabled, a single partition spec is derived from Delta's `partitionColumns` and column mapping field IDs.  All manifest entries reference this spec.

## Partition Values

> Partition values are stored in the `partition` struct (field ID 102) on each DATA entry. The struct schema is derived from the partition spec, using partition field IDs as struct field IDs.

Reviewer: So unlike the v4 stats, where the "user space" field ids are mapped to "metadata space" field ids, the partition value field ids from "user space" are directly nested inside the "metadata space" `partition` struct (with metadata space field ID 102)? How does that work?

When producing V4 manifest entries, writers convert Delta's `partitionValues` string map to the typed `partition` struct. When reading V4 entries back into Delta actions, readers extract partition values from the `partition` struct.

## Commit Types

Writers can produce two types of commits:

### Log Commit

A **log commit** writes changes directly to the Delta log JSON file without updating the V4 metadata tree. This is suitable for:
- Small appends (few files added)
- Metadata-only changes (schema updates, table properties)

> Log commits contain standard Delta actions (`add`, `remove`, `metadata`, etc.) without a `checkpoint` action.

Reviewer: I thought it was possible (and desirable) to do "checkpoint" commits that roll up all changes since the previous metadata tree change? Especially useful when the number of commits grows faster than the net changes? Would that be a manifest commit where the "new V4 metadata tree" is not actually new at all? Or is it a missing use case?

Author: To keep the implementations simple, should we consider just having manifest commits and log commits? ie the high churn case, we will just have a new tree that have minimal content changes than the previous tree. ie in the JSON logs, the checkpoint version is assumed to be same as the content root version. 

I'm open to the alternative though. ie the manifest commit has the freedom to just do a log compaction and not write the root. We could just allow this on the commit path as well

Writers should trigger manifest commits to limit the number of JSON log files and the volume of JSON bytes readers must consume. The specific thresholds are implementation-defined.

Manifest commits have the following characteristics:

1. **File actions may be logged**: A manifest commit may also
   write `add` and `remove` actions to the Delta log, in addition to updating the V4 tree. The `checkpoint.version` may be less than the commit version (the tree covers up to `checkpoint.version`; remaining changes are in the log). Checkpoint versions must be strictly monotonically increasing across all checkpoint actions in the log.

2. **Non-file actions must be logged**: A manifest commit must always
   write non-file actions (`metadata`, `protocol`, `txn`, `domainMetadata`, `commitInfo`) to the Delta log. These actions are not stored in the V4 tree.

3. **Incorporates preceding commits**: Manifest commits must
   incorporate all preceding log commits (since the last checkpoint) into the new V4 tree.

### Standalone Checkpoint

Independently of commits, any writer can produce a standalone checkpoint file that references an existing V4 tree without producing a new one.

> A standalone checkpoint uses the V2 checkpoint naming scheme (`n.checkpoint.u.parquet`) and contains a `checkpoint` action. It must reference an existing tree via `contentRoot` and must not produce a new root manifest. File actions (`add`, `remove`) since that tree's version are included inline in the checkpoint file (not in sidecars). This keeps standalone checkpoints cheap: any writer can produce one without the cost of building a new tree. The inline file actions are bounded: if the volume of pending file actions were large, the writer would produce a manifest commit instead.

Reviewer: If the checkpoint is to be useful as a compat bridge with unaware clients, it needs to contain at least P&M fields in the names/places legacy clients would expect to find them. Which would mean classic checkpoint naming schema must be supported (`n.checkpoint.parquet`) and the checkpoint action fields need to be exploded out as top-level columns.

## Manifest Deletion Vectors (MDVs)

MDVs reduce the write amplification: instead of rewriting a leaf manifest to remove a file, the root manifest records an MDV that marks row positions in the leaf manifest as deleted.

MDVs are stored in `manifest_info.dv` on `DATA_MANIFEST` entries in the root manifest. See [Manifest Deletion Vectors](#manifest-deletion-vectors-mdvs) for field definitions and CDF bitmap semantics.

Example root manifest entry with MDV:

```
DATA_MANIFEST entry:
  content_type: 3 (DATA_MANIFEST)
  location: "metadata/leaf-m1.parquet"
  tracking:
    status: EXISTING
  manifest_info:
    dv: <bitmap marking positions 2, 17>
    dv_cardinality: 2
```

When reading a leaf manifest, readers must check the `DATA_MANIFEST` entry for a non-null `manifest_info.dv`. If present, deserialize the bitmap and skip entries at positions marked in it.

## Reader Requirements

When `adaptiveMetadata` is supported and active, readers must:

1. **Find the latest checkpoint**: Scan the Delta log for the most recent

>    commit containing a `checkpoint` action, or use `_last_checkpoint` if available.

Reviewer: A normal `checkpoint.*.parquet` file would also suffice, no?

Author: Yes. I clarified this. PTAL.

2. **Read the content root**: Parse the V4 root manifest file referenced by
   `checkpoint.contentRoot.path`.

3. **Apply MDVs**: When reading leaf manifests, skip entries at positions
   marked in `manifest_info.dv`.

4. **Replay log commits**: Apply any log commits after `checkpoint.version`
   to get the current table state.

5. **Handle sidecars**: If `checkpoint.sidecars` is present, read auxiliary

>    data from the referenced sidecar files.

Reviewer: Most of this information seems to be misplaced, perhaps better to update https://github.com/delta-io/delta/blob/master/PROTOCOL.md#action-reconciliation section instead?

Reviewer: Similar critique applies to the writer requirements below. 

In general, it would be good to look at reader/writer requirement sections for other features and match what they present and how. Much of the information in these sections probably belongs elsewhere.

Author: I wasn't able to fully fold them into action reconciliation. I frame the reader/writer requirements in the same mold as the other RFCs that got merged into the protocol.  (e.g. row tracking)

6. **Read partition values**: Extract partition values from the

>    `partition` struct (field ID 102) on DATA entries.

Reviewer: This one seems a bit out of place? It's not related to log replay like the others?

In particular,. what would the reader _do_ with that information? (presumably use it to prune, but this section is about correctness, not pruning)

Author: Yes, it doesn't belong there. I removed it.

## Writer Requirements

When `adaptiveMetadata` is supported and active, writers must:

1. **Track backreferences**: Writers must record the manifest location
   (manifest path and row position) for every file they read from the tree. For `remove` actions, the backreference identifies the entry being deleted. For `add` actions that supersede an existing entry (stats backfill, DV update), the backreference identifies the entry being replaced. See [Backreferences](#backreferences).

2. **Choose commit type**: Based on operation size and accumulated log entries:
   - Small operations -> log commit
   - Large operations or compaction threshold reached -> manifest commit

3. **Maintain two-level hierarchy**: The metadata tree must have at most two levels
   (root -> leaves). Writers must not create nested manifest references.

4. **Incorporate concurrent commits**: When producing a manifest commit,

>    writers must incorporate any concurrent log commits that succeeded before the manifest commit.

Reviewer: What does that actually mean? Sounds awfully complicated for the spec to just mention in passing like this?

Author: That is fair feedback. That one line was compressing the whole optimistic-concurrency/retry story into a passing mention. It's actually spelled out in the Conflict Resolution section (Log + Manifest, etc.), so I've replaced the requirement with a pointer there rather than half-describing it inline.

| <ins>sidecars</ins> | <ins>Array[Struct]</ins> | <ins>Optional list of sidecar entries for overflow txns and domain metadata only. File-level metadata is stored in the content tree. Each entry contains path, sizeInBytes, modificationTime, and tags (matching the existing sidecar action schema).</ins> |

<ins>When `v4` is true, readers can use `contentRoot.path` to begin prefetching the root manifest immediately. The `protocol`, `metadata`, `domainMetadata`, `txns`, and `sidecars` fields provide the complete table state at `version`. If the hint is absent, stale, or does not contain V4 fields, readers fall back to log replay to locate the latest `checkpoint` action.</ins>

### Checkpoints

> ***Change to [existing section](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#checkpoints)***

<ins>When the `adaptiveMetadata` table feature is enabled, checkpoint information may be embedded directly in Delta log entries via the `checkpoint` action, rather than stored in separate checkpoint files. See [Checkpoint Action](#checkpoint-action) for details.</ins>

> <ins>The `checkpoint` action is self-contained: it embeds the content root reference along with all non-content metadata (`protocol`, `metaData`, `domainMetadata`, `txns`, `sidecars`). Together with the referenced manifest tree, a single `checkpoint` action represents the complete table state up to `checkpoint.version`. Log commits after that version must still be replayed.</ins>

Author: This means the JSON row can be pretty large. Is that OK? cc @scovich

Reviewer: Seems like we have three options:
1. One row containing a `checkpoint` action
2. Flatten the checkpoint action, producing several rows (but some of those rows are arrays of actions, e.g. `txns`)
3. Fully flatten the checkpoint action into something like today's JSON commit, where every row is a distinct action

Any reason not to do 3/?

Author: The reason why we cannot do option #3 is because if we would not be able to distinguish whether a row in the log is part of the checkpoint action or whether it's part of that commit itself. I am inclined to just have one row for the checkpoint action, even if it gets a little long.

Reviewer: I merely meant that the `checkpoint` action could be an array instead of an object. 
Not that its content would mingle with the normal rows of the containing commit.

Author: Thanks. I have changed it such that it is an array, and also clarified the different representation in standalone checkpoints.

The `_delta_log/_sidecars/` directory stores auxiliary data that is too large to embed in the checkpoint action.

## Checkpoint Action

When a manifest commit occurs, the Delta log entry contains a self-contained `checkpoint` action that includes all non-content metadata needed to reconstruct the full table state.

```json
{
  "checkpoint": {
    "version": 42,
    "contentRoot": {
      "path": "metadata/a3d1f7e2-v42.parquet",
      "sizeInBytes": 1024
    },
    "protocol": {
      "minReaderVersion": 3,
      "minWriterVersion": 7,
      "readerFeatures": ["columnMapping", "deletionVectors", "adaptiveMetadata"],
      "writerFeatures": ["columnMapping", "deletionVectors", "domainMetadata",
                         "rowTracking", "adaptiveMetadata"]
    },
    "metaData": {
      "id": "af23c9d7-fff1-4a5a-a2c8-55c59bd782aa",
      "name": "my_table",
      "schemaString": "{...}",
      "partitionColumns": [],
      "configuration": {},
      "createdTime": 1234567890000
    },

>     "domainMetadata": [

Reviewer 3: Should we break this into system vs other domain metadata?

system = starts with `delta.` .

We can suggest following:
- system domain metadata must be inlined
- other domain metadata can be inlined (if small) vs kept in sidecar.

If a checkpoint action has "other domain metadata" inlined as well as a sidecar corresponding to it -- its invalid. One and exactly one of these should be present.

Author: That makes sense to me. I have incorporated it.

Reviewer: Is this comment stale? I thought we had settled on the inline and sidecar merely being disjoint (no one action appears in both) rather than mutually exclusive (no one commit takes both approaches)?

Author: Yes, this is stale. This is what we have currently:

> System domain metadata (domains prefixed with `delta.`) must always appear as `domainMetadata` entries in the checkpoint action; it is small and required for correctness. User domain metadata and `txn`s may appear as entries, in sidecars, or split across both.

> For both user domain metadata and `txns`, inline and sidecar storage may coexist: a small inline set and one or more sidecars can be present together. The only constraint is per-key uniqueness — a given `appId` (or `domain`) must appear in exactly one place, whether inline or in a single sidecar, never in more than one.

>     "domainMetadata": [

Reviewer 3: For DomainMetadata, we use key `domainMetadata` (Without s).
For SetTransaction, we use key `txns` (with s).
For Sidecar, we use key `sidecars` (with s).

We should be consistent b/w all three.

Maybe we can skip `s` everywhere - No strong preference though.

Author: This is intentional, as the word `data` is considered to be plural as well.

      {
        "domain": "delta.rowTracking",
        "configuration": "{\"rowIdHighWaterMark\": 1000000}",
        "removed": false
      }
    ],

>     "txns": [

Reviewer 3: Same here - we should not allow txns inlined as well as txns present in sidecar.

Reviewer 3: I say following semantics makes everything explicit:

- `"txns": [ ]` ==> txns are inlined and they are empty. i.e. this table has no set transaction actions.

- `"txns": [ ... ]` ====> txns are inlined and they are non-empty. There must not be any sidecars of type = txn.

- `txns key missing` ====> txns are not inlined. There must be one sidecar corresponding to `type = txn`.

Author: I was thinking that we can allow both. If a commit adds a single `appId`, they should just be allowed to write it inline. But the restriction is that a given appId appears either in inline or the sidecars, but not both.

      {
        "appId": "streaming-app-1",
        "version": 100,
        "lastUpdated": 1234567890000
      }
    ],
    "sidecars": [
      {
        "path": "txn-v42.parquet",
        "sizeInBytes": 2048,
        "modificationTime": 1234567890000,

>         "tags": {}

Reviewer 3: We should have `type: txn` here? This will help us disambiguate sidecars belonging to txn vs sidecars belonging to `otherDomainMetadata` .

Reviewer 3: Also should we allow multiple sidecars for same type? Lets be explicit about it.
I think we can as the sidecars are flat and doesn't involve any added complexity.

Author: sounds good.  Updated.

      }
    ]
  }
}
```

### Checkpoint Action Fields

| Field | Type | Description |
|-------|------|-------------|
| `version` | Long | The table version up to which the checkpoint is complete. May be less than or equal to the commit version (e.g., commit v100 may checkpoint v50). Checkpoint versions must be strictly monotonically increasing across all checkpoint actions in the log (the next checkpoint must be for a version > 50). |
| `contentRoot` | Struct | Reference to the V4 root manifest file |
| `contentRoot.path` | String | Path to the root manifest, relative to the table root. May be an absolute URI for files outside the table root. |
| `contentRoot.sizeInBytes` | Long | Size of the root manifest file in bytes |
| `protocol` | Struct | The Protocol action at this checkpoint version |
| `metaData` | Struct | The Metadata action at this checkpoint version |
| `domainMetadata` | Array[Struct] | Optional. DomainMetadata actions at this checkpoint version |
| `txns` | Array[Struct] | Optional. SetTransaction actions (appId, version, lastUpdated). When large, stored in sidecars. |

> | `sidecars` | Array[Struct] | Optional. Sidecar file references for overflow txns and domain metadata only. File-level metadata is stored in the content tree. Schema: path, sizeInBytes, modificationTime, tags. |

Author: Consider adding `sidecarType`

For ADDED entries in leaf manifests, both are null and inherited from the `DATA_MANIFEST` entry in the root (see [Inheritance](#inheritance)). For EXISTING entries (e.g., after compaction), both are materialized.

On compaction, the output file is a new physical file at the compaction commit version, so all three file-level fields (`defaultRowCommitVersion`, `sequence_number`, `file_sequence_number`) are the compaction version. The original per-row values (`_row_id`, `_row_commit_version`) are materialized as data columns in the output Parquet file to preserve row-level tracking.

## Partition Specs

Iceberg uses **partition specs** to define how a table is partitioned. Each partition spec has a unique `spec_id` and describes the partition columns.

Delta defines partitioning via `partitionColumns` in the table metadata and does not support partition evolution. When `adaptiveMetadata` is enabled, a single partition spec is derived from Delta's `partitionColumns` and column mapping field IDs.  All manifest entries reference this spec.

## Partition Values

> Partition values are stored in the `partition` struct (field ID 102) on each DATA entry. The struct schema is derived from the partition spec, using partition field IDs as struct field IDs.

Author: Let me explain how Iceberg works with an example. If table has two columns, say  `region` (field 5, string), `date` (field 8, date); partition spec:

```
{ source-id: 5, field-id: 1000, name: region, transform: identity }
{ source-id: 8, field-id: 1001, name: date,   transform: identity }
```

Then struct 102's type is:

```
partition: struct<
  1000: region: string,
  1001: date:   date
>
```

Note that they have a partition ID assignment machinery since they support partition evolution. I think what makes sense for Delta is to support a deterministic ID assignment. ie the `i`-th partitionColumns entry deterministically gets field-id = `1000 + i`. 

I have updated the RFC to make this clear.

## Motivation

Current Delta checkpoints rewrite the entire table state on every checkpoint, regardless of how much actually changed. This makes metadata costs proportional to table size rather than operation size, and prevents caching since checkpoint files change completely between versions.

Delta also stores all of a commit's changes directly in the commit JSON file, regardless of how large the commit is. Compared to a columnar manifest, JSON compresses roughly an order of magnitude worse than Parquet, cannot be read or written in parallel, and causes high memory pressure at both read and write time.

The `adaptiveMetadata` feature addresses these limitations by adopting a two-level tree structure where:
- A **root manifest** serves as the entry point, containing references to
  leaf manifests as well as inline data file entries
- **Leaf manifests** contain data file entries and remain stable across operations
- **Manifest Deletion Vectors (MDVs)** mark rows in leaf manifests as deleted without rewriting
  them

This design enables:
- Small tables to store all file entries inline in the root manifest (no leaf manifests needed)

> - Small appends to write only to the Delta log, or at most a new root manifest

Reviewer 4: Naive question / clarification: Where else would they write to? Today, all appends write to the Delta log?

Author: We are saying that now writers have the freedom to directly make their changes on the tree (or continue to use logs as today).

- Commits that delete files can update MDVs instead of rewriting manifests
- Metadata changes proportional to operation size
- Stable leaf manifests that can be cached effectively

--------

# Changes to existing sections

### Add File

> ***Change to [existing section](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#add-file-and-remove-file)***

<ins>When the `adaptiveMetadata` table feature is enabled, the `add` action supports a `backReference` field:</ins>

| Field Name | Data Type | Description |
| - | - | - |
| <ins>backReference</ins> | <ins>Struct</ins> | <ins>Reference to the existing entry in the metadata tree that this add supersedes (e.g., stats backfill, DV update). Null when the add is for a genuinely new file not already in the tree. Contains `manifest` (String) and `pos` (Long). See [Backreferences](#backreferences).</ins> |

### Remove File

> ***Change to [existing section](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#remove-file)***

> <ins>When the `adaptiveMetadata` table feature is enabled, the `remove` action requires a `backReference` field and `deletionTimestamp` must be null:</ins>

Reviewer 4: Comma after `backreference field`?
It requires (a) a backreference field, and (b) it also requires that `deletionTimetstamp` must be null?

Author: Reworded such that we have three distinct requirements clearly. 

> When the `adaptiveMetadata` table feature is enabled, the `remove` action must include a `backReference`, set `extendedFileMetadata` to true, and have a null `deletionTimestamp`:

| Field Name | Data Type | Description |
| - | - | - |
| <ins>checkpoint</ins> | <ins>Struct</ins> | <ins>The `checkpoint` action for the latest manifest commit. Schema defined in [Checkpoint Action](#checkpoint-action).</ins> |

<ins>When the embedded `checkpoint` action is present, readers can use `checkpoint.contentRoot.path` to begin prefetching the root manifest immediately, and the action provides the complete table state at `checkpoint.version`. If the hint is absent, stale, or does not contain a `checkpoint` action, readers fall back to log replay to locate the latest `checkpoint` action.</ins>

### Checkpoints

> ***Change to [existing section](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#checkpoints)***

<ins>When the `adaptiveMetadata` table feature is enabled, checkpoint information may be embedded directly in Delta log entries via the `checkpoint` action, rather than stored in separate checkpoint files. See [Checkpoint Action](#checkpoint-action) for details.</ins>

> <ins>The `checkpoint` action is self-contained: it embeds the content root reference along with all non-content metadata (`protocol`, `metaData`, `domainMetadata`, `txns`, `sidecars`). Together with the referenced manifest tree, a single `checkpoint` action represents the complete table state up to `checkpoint.version`. Log commits after that version must still be replayed.</ins>

Reviewer 4: nit: I don't believe we have defined in this RFC what content vs non-content metadata is

Author: Good callout. added it to the terminology.

The `adaptiveMetadata` table feature is supported when:
- The table is on Reader Version 3 and Writer Version 7.
- The feature `adaptiveMetadata` exists in the table `protocol`'s `readerFeatures` and `writerFeatures`.

Required table features that must also be enabled:
- `columnMapping` (`id` mode): Stable column identification across schema evolution
- `rowTracking`: Manifest entries natively carry row-tracking fields (`first_row_id`, `sequence_number`); see [Row Tracking Compatibility](#row-tracking-compatibility)
- `domainMetadata`: Storing feature-specific metadata
- `deletionVectors`: Deletes are represented as deletion vectors, not by rewriting files
- `inCommitTimestamp`: Reliable timestamp-based metadata cleanup and time travel

## Storage Layout

When `adaptiveMetadata` is enabled, the table storage layout includes a `metadata/` directory alongside the standard `_delta_log/`:

```
table/
├── _delta_log/
│   ├── 00000000000000000042.json     # Commit with embedded checkpoint action
│   ├── 00000000000000000043.json     # Log commit (small changes only)
│   ├── _sidecars/
│   │   └── txn-v42.parquet           # Transaction identifiers sidecar
│   └── _last_checkpoint

> ├── metadata/                          # metadata tree

Reviewer 4: nit: `#` off-by-one

Author: Fixed.

| `contentRoot` | Struct | Reference to the root manifest file |
| `contentRoot.path` | String | Path to the root manifest, relative to the table root. May be an absolute URI for files outside the table root. |
| `contentRoot.sizeInBytes` | Long | Size of the root manifest file in bytes |
| `contentRoot.version` | Long | The table version whose content this root manifest reflects. Must be `<= checkpoint.version`. In a manifest commit the two are equal — all content up to `checkpoint.version` is folded into the tree. In a standalone checkpoint it may be less, with the gap covered by inline file actions. |
| `protocol` | Struct | The Protocol action at this checkpoint version |
| `metaData` | Struct | The Metadata action at this checkpoint version |
| `domainMetadata` | Array[Struct] | Optional. Inlined DomainMetadata actions at this checkpoint version. System domains (keys prefixed with `delta.`) must always be inlined here. User domains may be inlined, stored in sidecars, or split across both. |

> | `txns` | Array[Struct] | Optional. SetTransaction actions (appId, version, lastUpdated). May be inlined here, stored in sidecars, or split across both (see below). |

Reviewer 4: Would it be nice to be allowed to have _recent_ `setTransactions` here, but still sidecar old ones? For a streaming ingest pipeline, it can find recent appIds + versions here (which it is most likely to conflict with) yet can continue to store the 100K+ other SetTxns in the sidecar?

Reviewer: "or split across both" would seem to cover that?

Author: Ryan's above comment is correct.

> | `sidecars` | Array[Struct] | Optional. Sidecar file references for user domain metadata and txns only. File-level metadata is stored in the content tree. Schema: type (`txn` or `domainMetadata`), path, sizeInBytes, modificationTime, tags. |

Reviewer 4: nit: would `setTransaction` be better?

Author: We should be consistent with the current terminology [here](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#transaction-identifiers), which uses `txn`.

Independently of commits, any writer can produce a standalone checkpoint file that references an existing metadata tree without producing a new one.

> A standalone checkpoint uses the V2 checkpoint naming scheme (`n.checkpoint.u.parquet`) and contains a `checkpoint` action. It must reference an existing tree via `contentRoot` and must not produce a new root manifest, so its `contentRoot.version` is that existing tree's version and is less than `checkpointMetadata.version`. File actions (`add`, `remove`) after `contentRoot.version` (up to `checkpointMetadata.version`) are included inline in the checkpoint file (not in sidecars). This keeps standalone checkpoints cheap: any writer can produce one without the cost of building a new tree. The inline file actions are bounded: if the volume of pending file actions were large, the writer would produce a manifest commit instead.

Reviewer: Let's force a normal `n.checkpoint.parquet` -- v2 checkpoints have a conflicting definition for `sidecar` column that could cause confusion. We could even make AMT and v2 checkpoints mutually exclusive table features, since v2 adds nothing of value on top of AMT.
